src/borrow_tracker/tree_borrows/perms.rs

```diff
@@ -267,6 +267,19 @@ impl Permission {
         transition::perform_access(kind, rel_pos, old_state, protected)
             .map(|new_state| PermTransition { from: old_state, to: new_state })
     }
+
+    /// During a provenance GC, we want to compact the tree.
+    /// For this, we want to merge nodes upwards if they have a singleton parent.
+    /// But we need to be careful: If the parent is Frozen, and the child is Reserved,
+    /// we can not do such a merge. In general, such a merge is possible if the parent
+    /// is not protected, and if the child is reachable from the parent. This
+    /// corresponds to the `partial_cmp` order on permissions.
+    pub fn can_be_replaced_by_child(self, child: Self) -> bool {
+        match self.inner.partial_cmp(&child.inner) {
+            Some(Ordering::Less) | Some(Ordering::Equal) => true,
+            Some(Ordering::Greater) | None => false,
+        }
+    }
```

Comment:
I also don't think this function is complete. For once, if the parent is `ReservedIM`, and the child is `Reserved`, we can still replace the parent by the child, whereas `partial_cmp` says that they are incomparable.

Reply:
Note that the handling of `Reserved {conflicted: true}` is *very* subtle:
* Usually, it's not OK to replace a conflicted `Reserved` by an `Active`. But a `Reserved` never has `Active` children anyways (as all `Active` have `Active` parents), and further, since we know the parent is not protected, we know that the conflictedness does not matter anyway
* This exposes another way the method is incomplete: a conflicted `Reserved` parent can be replaced by a non-conflicted `Reserved` child, since the parent's conflictedness no longer matters, as the parent is not protected. But this is not how `partial_cmp` works.

(Note that I am highlighting all these cases to show that the reasoning here is rather subtle in many cases)

Reply:
`partial_cmp` on permissions denotes reachability, and is not by definition tied to the amount of UB that the permission can produce. I like the idea of `can_be_replaced_by_child` but I think the above implementation suggests that the replaceability derives from `partial_cmp` when it's actually very dependent on invariants of the tree. I think it should be decoupled from `partial_cmp` and hand-written to explicitly show all compactable cases with their justification.

Reply:
For example the fact that you can compact an `Active` parent with a `Frozen` child is entirely dependent on whether the `Active` is protected, so there are at minimum many unwritten assumptions in `can_be_replaced_by_child` about the fact that protected tags can't be GC'd and the valid parent-child combinations.

Reply:
Yes, we do not remove protected parents. Such parents should never be GC'd anyways since they still have a protector end write coming.

But I do think there's a deeper connection between reachability and replaceability: Permissions that come "later" always have "less" permissions than those that come earlier. It should always be sound to replace a node by one that's further ahead in the state machine (protectors notwithstanding). But I agree that it should probably be decoupled in code.